php/media/class-gallery.php

```diff
@@ -0,0 +1,323 @@
+	 *
+	 * @param Media $media
+	 */
+	public function __construct( Media $media ) {
+		$this->media           = $media;
+		$this->original_config = isset( $media->plugin->config['settings']['gallery'] ) ? $media->plugin->config['settings']['gallery'] : null;
+
+		if ( $this->gallery_enabled() ) {
+			$this->setup_hooks();
+		}
+	}
+
+	/**
+	 * Gets the gallery settings in the expected json format.
+	 *
+	 * @return array
+	 */
+	public function get_config() {
+		if ( count( $this->config ) ) {
+			return $this->config;
+		}
+
+		if ( ! count( $this->original_config ) ) {
+			return array();
+		}
+
+		$config        = $this->original_config;
+		$custom_config = $config['custom_settings'];
+
+		// unset things that don't need to be in the final json.
+		unset( $config['enable_gallery'], $config['custom_settings'] );
+
+		$config = $this->prepare_config( $config );
+		$config = Utils::expand_dot_notation( $config );
+		$config = Utils::array_filter_recursive(
+			$config,
+			function ( $item ) {
+				return ! empty( $item );
+			}
+		);
+
+		$config['cloudName']   = $this->media->plugin->components['connect']->get_cloud_name();
+		$config['container']   = '.woocommerce-product-gallery';
+		$config['mediaAssets'] = array();
+
+		if ( ! empty( $custom_config ) ) {
+			$custom_config = json_decode( $custom_config, true );
+
+			if ( ! empty( $custom_config ) ) {
+				$config = array_merge( $config, $custom_config );
+			}
+		}
+
+		$this->config = $config;
+
+		return $config;
+	}
+
+	/**
+	 * Convert an array's keys to camelCase and transform booleans.
+	 * This is used for Cloudinary's gallery widget lib.
+	 *
+	 * @param array $input The array input that will have its keys camelcase-d.
+	 *
+	 * @return array
+	 */
+	public function prepare_config( array $input ) {
+		foreach ( $input as $key => $val ) {
+			if ( 'on' === $val || 'off' === $val ) {
+				$val = 'on' === $val;
+			} elseif ( is_numeric( $val ) ) {
+				$val = (int) $val;
+			}
+
+			if ( 'none' !== $val ) {
+				$new_key           = lcfirst( implode( '', array_map( 'ucfirst', explode( '_', $key ) ) ) );
+				$input[ $new_key ] = $val;
+			}
+
+			unset( $input[ $key ] );
+		}
+
+		return $input;
+	}
+
+	/**
+	 * Returns JSON format of the current config.
+	 *
+	 * @return string|null
+	 */
+	public function get_json() {
+		return count( $this->get_config() ) ? wp_json_encode( $this->get_config() ) : null;
+	}
+
+	/**
+	 * Register frontend assets for the gallery.
+	 */
+	public function enqueue_gallery_library() {
+		wp_enqueue_script(
+			self::GALLERY_LIBRARY_HANDLE,
+			self::GALLERY_LIBRARY_URL,
+			array(),
+			$this->media->plugin->version,
+			true
+		);
+
+		$product = wc_get_product();
+		$config  = $this->get_config();
+
+		if ( $product ) {
+			$config['mediaAssets'] = $this->get_image_data( $product->get_gallery_image_ids() );
+		}
+
+		wp_localize_script(
+			self::GALLERY_LIBRARY_HANDLE,
+			'cloudinaryGallery',
+			array( 'config' => wp_json_encode( $config ) )
```

Comment:
do we need to encode the `$config`?

Reply:
Yes, otherwise using it as an array directly we lose the the value types and everything becomes a string. I need integers and booleans to remain in tact.

Reply:
@dugajean - We need to use `wp_add_inline_script` when adding code.

Reply:
@DavidCramer I wanna stay away from writing actual JS in PHP. This current way works really well. 😄

Reply:
@dugajean - sure it works, but this is a hack. We need to follow best practises which is to use the correct functions for the task.